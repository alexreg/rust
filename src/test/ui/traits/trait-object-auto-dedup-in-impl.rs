// ignore-tidy-linelength

// Checks to make sure that `dyn Trait + Send` and `dyn Trait + Send + Send` are the same type.
// Issue: #47010

struct Struct;

impl Trait for Struct {}

trait Trait {}

type Send1 = Trait + Send;
type Send2 = Trait + Send + Send;
//~^ WARNING duplicate auto trait `Send` found in type parameter bounds [duplicate_auto_traits_in_bounds]

fn main () {}

impl Trait + Send {
    fn test(&self) {
    //~^ ERROR duplicate definitions with name `test`
        println!("one");
    }
}

impl Trait + Send + Send {
//~^ WARNING duplicate auto trait `Send` found in type parameter bounds [duplicate_auto_traits_in_bounds]
    fn test(&self) {
        println!("two");
    }
}
