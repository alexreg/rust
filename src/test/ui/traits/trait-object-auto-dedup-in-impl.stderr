warning: duplicate auto trait `Send` found in type parameter bounds
  --> $DIR/trait-object-auto-dedup-in-impl.rs:13:22
   |
LL | type Send2 = Trait + Send + Send;
   |                      ^^^^   ^^^^ subsequent use of auto trait
   |                      |
   |                      first use of auto trait
   |
   = note: #[warn(duplicate_auto_traits_in_bounds)] on by default

warning: duplicate auto trait `Send` found in type parameter bounds
  --> $DIR/trait-object-auto-dedup-in-impl.rs:25:14
   |
LL | impl Trait + Send + Send {
   |              ^^^^   ^^^^ subsequent use of auto trait
   |              |
   |              first use of auto trait
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #56522 <https://github.com/rust-lang/rust/issues/56522>

error[E0592]: duplicate definitions with name `test`
  --> $DIR/trait-object-auto-dedup-in-impl.rs:20:5
   |
LL | /     fn test(&self) {
LL | |     //~^ ERROR duplicate definitions with name `test`
LL | |         println!("one");
LL | |     }
   | |_____^ duplicate definitions for `test`
...
LL | /     fn test(&self) {
LL | |         println!("two");
LL | |     }
   | |_____- other definition for `test`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0592`.
